helpers/handleRequest.js

```diff
@@ -1,94 +1,93 @@
-	else shouldCache = reqShouldCache;
-	let cacheResult = false;
-	const hashAttribute = md5(`${method}${url.href}`);
-	const source = request.hostname;
-
-	if (shouldCache) {
-		cacheResult = await makeCacheRequest({
-			hdbCore,
-			hashAttribute,
-			minCreatedDate,
-		});
+import config from './config.js';
+import { md5 } from './utils.js';
+import { getRecord, saveRecord, updatingRecord } from './cacheHelper.js';
+import { callAPI } from './callAPI.js';
+
+export const getResults = async ({ hdbCore, logger, method, url, headers, body, shouldCache = true }) => {
+	// 1. parse url into an md5 id
+	// 2. check for shouldSkipCache
+	//		a. call API
+	//		b. return results
+	// 3. check for existing record
+	//		a. return 304 if exists and updatedtime < headers[if-modified-since]
+	// 4. check if record is updating
+	//		a. wait for update to complete
+	// 5. check if record is out of date (consult headers[access-control-max-age])
+	//		a. set record to updating
+	//		b. call API for new data
+	//		c. save new record
+	//		d. get record
+	// 6. return the cached record
+
+	// 1. CREATE ID from the method, url, and any headers (specified in the config)
+	let id = `${method}${url}`;
+	for (let header in config.HEADERS_TO_CACHE) {
+		if (header in headers) {
+			id += `:${header}=${headers[header]}`;
+		}
 	}
+	id = md5(id); // because a string with a forward slash cannot be a record id
 
-	if (cacheResult?.length) {
-		reply.header('hdb-from-cache', true);
-
-		const { result, content_type, status } = cacheResult[0];
-
-		reply.header('content-type', content_type);
-
-		return { status_code: status, response_body: result };
-	} else {
-		reply.header('hdb-from-cache', false);
-
-		try {
-			let recordExists = false;
-			if (shouldCache) {
-				const result = await setRecordToUpdating({ hdbCore, hashAttribute });
-				recordExists = result.update_hashes.length > 0;
-			}
+	// 2. SKIP CACHE
+	if (!config.METHODS_TO_CACHE.includes(method)) {
+		shouldCache = false;
+	}
 
-			const response = await callAPI({ request, url, headers });
-			const { content_type } = response;
-			const result = response.body;
-			const status = response.statusCode;
-			const error = response.statusCode < 200 || response.statusCode > 299;
-			const duration_ms = Date.now() - start;
+	// if not caching, get and return the data
+	if (!shouldCache) {
+		const { body: replyBody, status, headers: apiHeaders } = await callAPI({ request, url, headers });
+		const replyHeaders = Object.assign({}, apiHeaders, { 'hdb-from-cache': false });
+		return { status, replyBody, replyHeaders };
+	}
 
-			reply.header('content-type', content_type);
+	// 3. GET RECORD
+	let cachedRecord = await getRecord({ hdbCore, id });
+	if (cachedRecord && headers['if-modified-since']) {
+		const ifModifiedSince = new Date(headers['if-modified-since']);
+		const updatedTime = new Date(cachedRecord.__updatedtime__);
+		if (updatedTime < ifModifiedSince) {
```

Comment:
I think this should be `<=`. If they match (which should be the case if the client's cache is fresh), the dates should be equal.  Also, fun fact: you don't have to convert `__updatedtime__` to a `Date`, the comparison will coerce it (not that you should actually do that trick).

Reply:
That's cool. Is it because it's a timestamp already, and Date just turns into a timestamp whenever it's being compared?
Updated the less than to less than or equal to.

Reply:
Yeah, that's how Date comparison actually work in JS, JS coerces the dates to numbers (epoch ms) and then does the comparisons (because JS doesn't have operator overloading that might permit Date to define its own comparison, like is available in some languages).